language-plutus-core/src/Language/PlutusCore/CkMachine.hs

```diff
@@ -0,0 +1,162 @@
+{-# LANGUAGE ConstraintKinds      #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Language.PlutusCore.CkMachine ( CkError(..)
+                                     , CkException(..)
+                                     , evaluateCk
+                                     ) where
+
+import           PlutusPrelude
+import           Language.PlutusCore.Type
+import           Language.PlutusCore.Constant
+
+infix 4 |>, <|
+
+data Frame tyname name a = FrameApplyFun (Term tyname name a)
+                         | FrameApplyArg (Term tyname name a)
+                         | FrameTyInstArg
+                         | FrameUnwrap
+                         | FrameWrap a (tyname a) (Type tyname a)
+
+type Context tyname name a = [Frame tyname name a]
+
+data CkError = NonConstantReturnedCkError
+             | NonTyAbsInstantiatedError
+             | NonWrapUnwrappedCkError
+             | NonPrimitiveApplicationCkError
+             | OpenTermEvaluatedCkError
+
+data CkException tyname name = CkException
+    { _ckExceptionError :: CkError
+    , _ckExceptionCause :: Term tyname name ()
+    }
+
+data CkEvaluationResult = CkEvaluationSuccess (Constant ())
+                        | CkEvaluationFailure
+
+ckErrorString :: CkError -> String
+ckErrorString NonConstantReturnedCkError     =
+    "returned a non-constant"
+ckErrorString NonTyAbsInstantiatedError      =
+    "attempted to reduce a non-type-abstraction applied to a type"
+ckErrorString NonWrapUnwrappedCkError        =
+    "attempted to unwrap a not wrapped term"
+ckErrorString NonPrimitiveApplicationCkError =
+    "attempted to reduce a not immediately reducible application"
+ckErrorString OpenTermEvaluatedCkError       =
+    "attempted to evaluate an open term"
+
+instance Pretty (Term tyname name ()) => Show (CkException tyname name) where
+    show (CkException err cause) = concat
+        ["The CK machine " , ckErrorString err , ": " , prettyString cause]
+
+instance (Pretty (Term tyname name ()), Typeable tyname, Typeable name) =>
+    Exception (CkException tyname name)
+
+-- | Check whether a term is a value.
+isValue :: Term tyname name a -> Bool
+isValue (TyAbs  _ _ _ body) = isValue body
+isValue (Wrap   _ _ _ term) = isValue term
+isValue (LamAbs _ _ _ body) = isValue body
+isValue (Constant _ _)      = True
+isValue _                   = False
+
+-- | Substitute a term for a variable in a term that can contain duplicate binders.
+-- Do not descend under binders that bind the same variable as the one we're substituting for.
+substituteDb
+    :: Eq (name a)
+    => name a -> Term tyname name a -> Term tyname name a -> Term tyname name a
+substituteDb varFor new = go where
+    go (Var ann var)            = if var == varFor then new else Var ann var
+    go (TyAbs ann tyn ty body)  = TyAbs ann tyn ty (go body)
+    go (LamAbs ann var ty body) = LamAbs ann var ty (goUnder var body)
+    go (Apply ann fun arg)      = Apply ann (go fun) (undefined (go (undefined arg)))
```

Comment:
What is the purpose of the `undefined`s here?

Reply:
All those `undefined` are due to `arg` being a spine rather than a single argument. So I just pretend `arg` is a single argument. This will be fixed once `Apply` is binary function application.